src/dev/code_coverage/ingest_coverage/transforms.js

```diff
@@ -0,0 +1,165 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { left, right } from './either';
+import { always, id, noop } from './utils';
+
+const maybeTotal = x => (x === 'total' ? left(x) : right(x));
+
+export const trimLeftFrom = (text, x) => x.replace(new RegExp(`(?:.*)(${text}.*$)`, 'gm'), '$1');
+
+export const statsAndstaticSiteUrl = (...xs) => {
+  const [staticSiteUrl] = xs[0][1];
+  const [stats] = xs[0];
+  return {
+    staticSiteUrl,
+    ...stats,
+  };
+};
+
+export const addJsonSummaryPath = jsonSummaryPath => obj => ({
+  jsonSummaryPath: trimLeftFrom('target', jsonSummaryPath),
+  ...obj,
+});
+
+export const truncate = text => obj => {
+  const { staticSiteUrl } = obj;
+  if (staticSiteUrl.includes(text)) obj.staticSiteUrl = trimLeftFrom(text, staticSiteUrl);
+  return obj;
+};
+
+export const addTimeStamp = ts => obj => ({
+  ...obj,
+  '@timestamp': ts,
+});
+
+const splitOnRoot = root => x => x.split(root)[1];
+const setTotal = x => obj => (obj.isTotal = x);
+const mutateTrue = setTotal(true);
+const mutateFalse = setTotal(false);
+
+const root = urlBase => ts => testRunnerType =>
+  `${urlBase}/${ts}/${testRunnerType.toLowerCase()}-combined`;
+
+const prokForTotalsIndex = mutateTrue => urlRoot => obj =>
+  right(obj)
+    .map(mutateTrue)
+    .map(always(`${urlRoot}/index.html`))
+    .fold(noop, id);
+
+const prokForCoverageIndex = root => mutateFalse => urlRoot => obj => siteUrl =>
+  right(siteUrl)
+    .map(x => {
+      mutateFalse(obj);
+      return x;
+    })
+    .map(x => {
+      return splitOnRoot(root)(x);
+    })
+    .map(x => `${urlRoot}${x}.html`)
+    .fold(noop, id);
+
+export const staticSite = urlBase => obj => {
+  const { staticSiteUrl, testRunnerType, COVERAGE_INGESTION_KIBANA_ROOT } = obj;
+  const ts = obj['@timestamp'];
+  const urlRoot = root(urlBase)(ts)(testRunnerType);
+  const prokTotal = prokForTotalsIndex(mutateTrue)(urlRoot);
+  const prokCoverage = prokForCoverageIndex(COVERAGE_INGESTION_KIBANA_ROOT)(mutateFalse)(urlRoot)(
+    obj
+  );
+  const prokForBoth = always(maybeTotal(staticSiteUrl).fold(always(prokTotal(obj)), prokCoverage));
+
+  return { ...obj, staticSiteUrl: prokForBoth() };
+};
+
+export const coveredFilePath = obj => {
+  const { staticSiteUrl, COVERAGE_INGESTION_KIBANA_ROOT } = obj;
+
+  const withoutCoveredFilePath = always(obj);
+  const leadingSlashRe = /^\//;
+  const maybeDropLeadingSlash = x => (leadingSlashRe.test(x) ? right(x) : left(x));
+  const dropLeadingSlash = x => x.replace(leadingSlashRe, '');
+  const dropRoot = root => x =>
+    maybeDropLeadingSlash(splitOnRoot(root)(x)).fold(id, dropLeadingSlash);
+  return maybeTotal(staticSiteUrl)
+    .map(dropRoot(COVERAGE_INGESTION_KIBANA_ROOT))
+    .fold(withoutCoveredFilePath, coveredFilePath => ({ ...obj, coveredFilePath }));
+};
+
+export const ciRunUrl = obj => {
+  const ciRunUrl = process.env.CI_RUN_URL || 'CI RUN URL NOT PROVIDED';
```

Comment:
Should probably throw an error instead of setting a error message to the expected URL variable.

Reply:
Well, it's not absolutely needed, yet anyway.  You prefer to throw?

Reply:
One reason I'd prefer not to throw as this datum is not mission critical.  It takes over 2 hours to get here in ci.  We'd rather not throw

Reply:
That's fine, then we should find another way not to fail if it's not set. We should never be setting a URL variable to something other than a URL, it's just asking for trouble.

Reply:
Ahh, I we could truncate to just the ci url, minus the build number! Whatcha think?

Reply:
I think that's misusing a mechanism to control flow which it's not meant to do. Can we just avoid triggering the code coverage ingest if CI_RUN_URL is not set?

Reply:
Asking my boss now

Reply:
Yep, I think if we don't have a CI_RUN_URL (like a local run) we just don't ingest that field.

Reply:
I'll code that